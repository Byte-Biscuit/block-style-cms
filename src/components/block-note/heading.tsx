import React from "react";
import { twColorMap, type BlockData } from "./meta";
import Content from "./item/content";
import { getDefaultHeadingClasses } from "@/lib/tw-utils";

interface HeadingProps {
    data: BlockData & { props: { level: number } };
    className?: string;
}

export const Heading: React.FC<HeadingProps> = ({ data, className = "" }) => {
    const { props, content } = data;

    if (!props) return null;
    if (!content || content.length === 0) return null;

    const HeadingTag = `h${props.level}` as keyof React.JSX.IntrinsicElements;

    // build tag classes
    const tagClasses = [
        // tx color
        props.textColor !== "default" &&
            props.textColor &&
            twColorMap[props.textColor]?.textClass,
        // bg color
        props.backgroundColor !== "default" &&
            props.backgroundColor &&
            twColorMap[props.backgroundColor]?.bgClass,
        // textAlignment
        props.textAlignment === "center" && "text-center",
        props.textAlignment === "right" && "text-right",
        // Unify title style
        getDefaultHeadingClasses(props.level),
        // custom class name
        className,
    ]
        .filter(Boolean)
        .join(" ");
    return React.createElement(
        HeadingTag,
        { className: tagClasses },
        <Content items={content} />
    );
};

export default Heading;
