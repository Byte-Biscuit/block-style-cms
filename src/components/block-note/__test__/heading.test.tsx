import React from "react";
import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { Heading } from "../heading";
import type { BlockData, TextContent, LinkContent } from "../meta";

describe("Heading", () => {
    it("renders h3 heading with mixed text and link content", () => {
        const headingData: BlockData & { props: { level: number } } = {
            id: "c33232af-a262-468e-a7e9-4df7d899af56",
            type: "heading",
            props: {
                textColor: "red",
                backgroundColor: "green",
                textAlignment: "center",
                level: 3,
                isToggleable: false,
            },
            content: [
                {
                    type: "text",
                    text: "这是",
                    styles: {},
                },
                {
                    type: "link",
                    href: "https://studio.xiyue.space",
                    content: [
                        {
                            type: "text",
                            text: "文",
                            styles: {},
                        },
                        {
                            type: "text",
                            text: "字",
                            styles: {
                                bold: true,
                                italic: true,
                                underline: true,
                                strike: true,
                                textColor: "red",
                            },
                        },
                    ],
                },
                {
                    type: "text",
                    text: "数据",
                    styles: {},
                },
            ] as (TextContent | LinkContent)[],
            children: [],
        };

        const { asFragment } = render(<Heading data={headingData} />);

        // 验证渲染的是 h3 标签
        const headingElement = screen.getByRole("heading", { level: 3 });
        expect(headingElement).toBeTruthy();
        expect(headingElement.tagName).toBe("H3");

        // 验证链接元素
        const linkEl = screen.getByRole("link") as HTMLAnchorElement;
        expect(linkEl).toBeTruthy();
        expect(linkEl.href).toBe("https://studio.xiyue.space/");

        // 验证链接内的第二个文本 "字" (有样式)
        const linkText2 = screen.getByText("字");
        expect(linkText2).toBeTruthy();
        expect(linkText2.tagName).toBe("SPAN");

        const styledClassName = linkText2.className;
        expect(styledClassName).toContain("font-bold");
        expect(styledClassName).toContain("italic");
        expect(styledClassName).toContain("underline");
        expect(styledClassName).toContain("line-through");
        expect(styledClassName).toContain("text-red-600");

        // 验证标题包含所有文本内容 (使用 DOM 查询验证包含关系)
        expect(headingElement.contains(linkEl)).toBe(true);

        // 验证默认样式类（左对齐，默认颜色）
        const headingClassName = headingElement.className;
        expect(headingClassName).not.toContain("text-right");
        expect(headingClassName).not.toContain("text-left");

        // snapshot 验证
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders different heading levels correctly", () => {
        const h1Data = {
            id: "test-h1",
            type: "heading",
            props: { level: 1 },
            content: [
                { type: "text" as const, text: "H1 Title", styles: {} },
            ] as (TextContent | LinkContent)[],
            children: [],
        } as BlockData & { props: { level: number } };

        const { rerender } = render(<Heading data={h1Data} />);

        const h1Element = screen.getByRole("heading", { level: 1 });
        expect(h1Element.tagName).toBe("H1");

        // 测试 h6
        const h6Data = {
            ...h1Data,
            props: { level: 6 },
            content: [
                { type: "text" as const, text: "H6 Title", styles: {} },
            ] as (TextContent | LinkContent)[],
        };
        rerender(<Heading data={h6Data} />);

        const h6Element = screen.getByRole("heading", { level: 6 });
        expect(h6Element.tagName).toBe("H6");
    });

    it("returns null when props is missing", () => {
        const invalidData = {
            id: "invalid",
            type: "heading",
            content: [{ type: "text" as const, text: "Test", styles: {} }] as (
                | TextContent
                | LinkContent
            )[],
            children: [],
        } as unknown as BlockData & { props: { level: number } };

        const { container } = render(<Heading data={invalidData} />);
        expect(container.firstChild).toBeNull();
    });

    it("returns null when content is empty", () => {
        const emptyContentData = {
            id: "empty",
            type: "heading",
            props: { level: 2 },
            content: [],
            children: [],
        } as BlockData & { props: { level: number } };

        const { container } = render(<Heading data={emptyContentData} />);
        expect(container.firstChild).toBeNull();
    });
});
