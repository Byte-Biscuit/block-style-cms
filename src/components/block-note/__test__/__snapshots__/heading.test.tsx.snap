// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Heading > renders h3 heading with mixed text and link content 1`] = `
<DocumentFragment>
  <h3
    class="text-red-600 bg-teal-100 text-center text-xl leading-7 sm:text-2xl sm:leading-8 md:text-3xl md:leading-9 font-bold tracking-tight"
  >
    这是
    <a
      class="break-words"
      href="https://studio.xiyue.space"
      rel="noopener noreferrer"
      target="_blank"
    >
      文
      <span
        class="font-bold italic underline line-through text-red-600"
      >
        字
      </span>
    </a>
    数据
  </h3>
</DocumentFragment>
`;
